File: bulk-create-users.js
#!/usr/bin/env node

/**
 * Bulk User Creation Script for eBay Analytics Platform
 * 
 * This script creates multiple user accounts with a default password.
 * 
 * Usage:
 *   node bulk-create-users.js emails.txt
 *   OR
 *   node bulk-create-users.js (will prompt for emails)
 * 
 * Requirements:
 *   - Node.js
 *   - Supabase Admin API key
 *   - npm install @supabase/supabase-js
 */

const { createClient } = require('@supabase/supabase-js');
const fs = require('fs');
const readline = require('readline');

// ============================================
// CONFIGURATION
// ============================================
const SUPABASE_URL = process.env.SUPABASE_URL || 'YOUR_SUPABASE_URL';
const SUPABASE_SERVICE_ROLE_KEY = process.env.SUPABASE_SERVICE_ROLE_KEY || 'YOUR_SERVICE_ROLE_KEY';
const DEFAULT_PASSWORD = 'Recharge';
const DEFAULT_FULL_NAME = 'User'; // Can be customized per user if needed

// ============================================
// INITIALIZE SUPABASE CLIENT (Admin)
// ============================================
const supabase = createClient(SUPABASE_URL, SUPABASE_SERVICE_ROLE_KEY, {
    auth: {
        autoRefreshToken: false,
        persistSession: false
    }
});

// ============================================
// HELPER FUNCTIONS
// ============================================

function validateEmail(email) {
    const emailRegex = /^[^\s@]+@[^\s@]+\.[^\s@]+$/;
    return emailRegex.test(email);
}

async function createUser(email, fullName = DEFAULT_FULL_NAME) {
    try {
        // Create auth user
        const { data: authData, error: authError } = await supabase.auth.admin.createUser({
            email: email,
            password: DEFAULT_PASSWORD,
            email_confirm: true, // Auto-confirm email
            user_metadata: {
                full_name: fullName
            }
        });

        if (authError) {
            return {
                success: false,
                email: email,
                error: authError.message
            };
        }

        // Create profile with trial period
        const trialEndsAt = new Date();
        trialEndsAt.setDate(trialEndsAt.getDate() + 14);

        const { error: profileError } = await supabase
            .from('profiles')
            .insert({
                id: authData.user.id,
                email: email,
                full_name: fullName,
                status: 'pending', // Users start as pending (need approval)
                subscription_status: 'trial',
                subscription_plan: 'free',
                trial_ends_at: trialEndsAt.toISOString()
            });

        if (profileError) {
            console.error(`⚠️ Profile creation failed for ${email}:`, profileError.message);
            // User was created but profile failed - might need manual fix
            return {
                success: false,
                email: email,
                error: `Profile creation failed: ${profileError.message}`
            };
        }

        return {
            success: true,
            email: email,
            userId: authData.user.id
        };
    } catch (error) {
        return {
            success: false,
            email: email,
            error: error.message
        };
    }
}

async function processEmails(emails) {
    const results = {
        total: emails.length,
        successful: [],
        failed: []
    };

    console.log(`\n🚀 Starting bulk user creation for ${emails.length} email(s)...\n`);

    for (let i = 0; i < emails.length; i++) {
        const email = emails[i].trim();
        
        if (!email) continue;
        
        if (!validateEmail(email)) {
            console.log(`❌ [${i + 1}/${emails.length}] Invalid email format: ${email}`);
            results.failed.push({ email, error: 'Invalid email format' });
            continue;
        }

        console.log(`📧 [${i + 1}/${emails.length}] Creating account for: ${email}...`);
        
        const result = await createUser(email);
        
        if (result.success) {
            console.log(`✅ Success: ${email} (User ID: ${result.userId})`);
            results.successful.push(result);
        } else {
            console.log(`❌ Failed: ${email} - ${result.error}`);
            results.failed.push(result);
        }

        // Small delay to avoid rate limiting
        if (i < emails.length - 1) {
            await new Promise(resolve => setTimeout(resolve, 200));
        }
    }

    return results;
}

function printSummary(results) {
    console.log('\n' + '='.repeat(60));
    console.log('📊 BULK USER CREATION SUMMARY');
    console.log('='.repeat(60));
    console.log(`Total emails: ${results.total}`);
    console.log(`✅ Successful: ${results.successful.length}`);
    console.log(`❌ Failed: ${results.failed.length}`);
    
    if (results.successful.length > 0) {
        console.log('\n✅ SUCCESSFUL CREATIONS:');
        results.successful.forEach(result => {
            console.log(`   - ${result.email}`);
        });
    }
    
    if (results.failed.length > 0) {
        console.log('\n❌ FAILED CREATIONS:');
        results.failed.forEach(result => {
            console.log(`   - ${result.email}: ${result.error}`);
        });
    }
    
    console.log('\n' + '='.repeat(60));
    console.log(`\n📝 Default password for all accounts: "${DEFAULT_PASSWORD}"`);
    console.log(`📝 Note: Users are created with status "pending" and need approval.\n`);
}

// ============================================
// MAIN SCRIPT
// ============================================

async function main() {
    let emails = [];

    // Check if email file is provided as argument
    if (process.argv[2]) {
        const filePath = process.argv[2];
        try {
            const fileContent = fs.readFileSync(filePath, 'utf-8');
            emails = fileContent
                .split('\n')
                .map(line => line.trim())
                .filter(line => line.length > 0);
            console.log(`📄 Loaded ${emails.length} email(s) from ${filePath}`);
        } catch (error) {
            console.error(`❌ Error reading file ${filePath}:`, error.message);
            process.exit(1);
        }
    } else {
        // Interactive mode - prompt for emails
        console.log('📧 Bulk User Creation Tool');
        console.log('Enter email addresses (one per line, or comma-separated):');
        console.log('Press Ctrl+D (Mac/Linux) or Ctrl+Z (Windows) when done, or enter "done"\n');
        
        const rl = readline.createInterface({
            input: process.stdin,
            output: process.stdout
        });

        const lines = [];
        rl.on('line', (line) => {
            if (line.trim().toLowerCase() === 'done') {
                rl.close();
            } else {
                lines.push(line);
            }
        });

        rl.on('close', async () => {
            // Parse emails (support both line-by-line and comma-separated)
            emails = lines
                .join('\n')
                .split(/[,\n]/)
                .map(email => email.trim())
                .filter(email => email.length > 0);
            
            if (emails.length === 0) {
                console.log('❌ No emails provided. Exiting.');
                process.exit(1);
            }

            // Validate configuration
            if (SUPABASE_URL === 'YOUR_SUPABASE_URL' || SUPABASE_SERVICE_ROLE_KEY === 'YOUR_SERVICE_ROLE_KEY') {
                console.error('\n❌ ERROR: Please configure SUPABASE_URL and SUPABASE_SERVICE_ROLE_KEY');
                console.error('   Set them as environment variables or edit this script.\n');
                process.exit(1);
            }

            const results = await processEmails(emails);
            printSummary(results);
            process.exit(results.failed.length > 0 ? 1 : 0);
        });

        return;
    }

    // Validate configuration
    if (SUPABASE_URL === 'YOUR_SUPABASE_URL' || SUPABASE_SERVICE_ROLE_KEY === 'YOUR_SERVICE_ROLE_KEY') {
        console.error('\n❌ ERROR: Please configure SUPABASE_URL and SUPABASE_SERVICE_ROLE_KEY');
        console.error('   Set them as environment variables or edit this script.');
        console.error('   Example: SUPABASE_URL=https://xxx.supabase.co SUPABASE_SERVICE_ROLE_KEY=xxx node bulk-create-users.js emails.txt\n');
        process.exit(1);
    }

    if (emails.length === 0) {
        console.error('❌ No valid emails found. Exiting.');
        process.exit(1);
    }

    const results = await processEmails(emails);
    printSummary(results);
    process.exit(results.failed.length > 0 ? 1 : 0);
}

// Run the script
main().catch(error => {
    console.error('❌ Fatal error:', error);
    process.exit(1);
});

File: bulk-create-users.py
#!/usr/bin/env python3
"""
Bulk User Creation Script for eBay Analytics Platform (Python Version)

This script creates multiple user accounts with a default password.

Usage:
    python3 bulk-create-users.py emails.txt
    
    OR
    
    SUPABASE_URL=https://xxx SUPABASE_SERVICE_ROLE_KEY=xxx python3 bulk-create-users.py emails.txt

Requirements:
    - Python 3.6+
    - requests library
    - Supabase Admin API key
"""

import os
import sys
import json
import time
from datetime import datetime, timedelta

try:
    import requests
except ImportError:
    print("❌ Error: 'requests' library not installed")
    print("📦 Install it with: pip install requests")
    sys.exit(1)

# ============================================
# CONFIGURATION
# ============================================
SUPABASE_URL = os.environ.get('SUPABASE_URL', 'https://yknvgrydvxnkzycpjblv.supabase.co')
SUPABASE_SERVICE_ROLE_KEY = os.environ.get('SUPABASE_SERVICE_ROLE_KEY', 'YOUR_SERVICE_ROLE_KEY')
DEFAULT_PASSWORD = 'Recharge'

# ============================================
# VALIDATION
# ============================================
def validate_email(email):
    """Validate email format"""
    email = email.strip()
    if '@' not in email or '.' not in email:
        return False
    return True

def validate_config():
    """Validate configuration"""
    if 'YOUR_' in SUPABASE_URL or 'YOUR_' in SUPABASE_SERVICE_ROLE_KEY:
        print("❌ Error: Please configure SUPABASE_URL and SUPABASE_SERVICE_ROLE_KEY")
        print("\nYou can set them as environment variables:")
        print("  export SUPABASE_URL=https://your-project.supabase.co")
        print("  export SUPABASE_SERVICE_ROLE_KEY=your-service-role-key")
        print("\nOr run with one-line command:")
        print("  SUPABASE_URL=xxx SUPABASE_SERVICE_ROLE_KEY=xxx python3 bulk-create-users.py emails.txt")
        print("\nGet your service role key from:")
        print("  https://app.supabase.com/project/yknvgrydvxnkzycpjblv/settings/api")
        return False
    return True

# ============================================
# SUPABASE API FUNCTIONS
# ============================================

def create_auth_user(email, password):
    """Create user in Supabase Auth"""
    url = f"{SUPABASE_URL}/auth/v1/admin/users"
    headers = {
        'apikey': SUPABASE_SERVICE_ROLE_KEY,
        'Authorization': f'Bearer {SUPABASE_SERVICE_ROLE_KEY}',
        'Content-Type': 'application/json'
    }
    
    data = {
        'email': email,
        'password': password,
        'email_confirm': True,
        'auto_confirm_user': True
    }
    
    try:
        response = requests.post(url, headers=headers, json=data, timeout=30)
        if response.status_code in [200, 201]:
            return {'success': True, 'user': response.json()}
        else:
            error_msg = response.text
            if 'already registered' in error_msg or 'already exists' in error_msg or 'duplicate' in error_msg.lower():
                return {'success': False, 'error': 'User already registered'}
            return {'success': False, 'error': error_msg[:200]}
    except Exception as e:
        return {'success': False, 'error': str(e)}

def create_profile(user_id, email):
    """Create profile in profiles table"""
    url = f"{SUPABASE_URL}/rest/v1/profiles"
    headers = {
        'apikey': SUPABASE_SERVICE_ROLE_KEY,
        'Authorization': f'Bearer {SUPABASE_SERVICE_ROLE_KEY}',
        'Content-Type': 'application/json',
        'Prefer': 'return=representation'
    }
    
    # Calculate 14 days from now
    trial_end = (datetime.now() + timedelta(days=14)).isoformat()
    
    data = {
        'id': user_id,
        'email': email,
        'full_name': email.split('@')[0].title(),  # Use email prefix as name
        'status': 'pending',
        'subscription_status': 'trial',
        'subscription_plan': 'free',
        'trial_ends_at': trial_end,
        'created_at': datetime.now().isoformat(),
        'updated_at': datetime.now().isoformat()
    }
    
    try:
        response = requests.post(url, headers=headers, json=data, timeout=30)
        if response.status_code in [200, 201]:
            return {'success': True, 'profile': response.json()}
        else:
            return {'success': False, 'error': response.text[:200]}
    except Exception as e:
        return {'success': False, 'error': str(e)}

def create_user(email):
    """Create complete user account"""
    # Validate email
    if not validate_email(email):
        return {'success': False, 'error': 'Invalid email format'}
    
    # Create auth user
    auth_result = create_auth_user(email, DEFAULT_PASSWORD)
    if not auth_result['success']:
        return auth_result
    
    user_id = auth_result['user']['id']
    
    # Create profile
    profile_result = create_profile(user_id, email)
    if not profile_result['success']:
        # Auth user created but profile failed
        print(f"   ⚠️  Warning: Profile creation failed for {email}, but user was created")
        return {'success': True, 'user_id': user_id, 'profile_warning': True}
    
    return {'success': True, 'user_id': user_id, 'profile': profile_result['profile']}

# ============================================
# MAIN EXECUTION
# ============================================

def main():
    print("🚀 eBay Analytics - Bulk User Creation")
    print("=" * 60)
    
    # Validate configuration
    if not validate_config():
        sys.exit(1)
    
    # Get email list
    email_list = []
    if len(sys.argv) > 1:
        # Read from file
        filename = sys.argv[1]
        try:
            with open(filename, 'r') as f:
                email_list = [line.strip() for line in f if line.strip() and not line.startswith('#')]
        except FileNotFoundError:
            print(f"❌ Error: File '{filename}' not found")
            sys.exit(1)
    else:
        print("\nPlease enter email addresses (one per line, or comma-separated).")
        print("Type 'done' when finished:\n")
        while True:
            try:
                line = input().strip()
                if line.lower() == 'done':
                    break
                emails = [e.strip() for e in line.split(',')]
                email_list.extend(emails)
            except EOFError:
                break
    
    if not email_list:
        print("❌ No email addresses provided")
        sys.exit(1)
    
    # Remove duplicates
    email_list = list(dict.fromkeys(email_list))
    
    print(f"\n🚀 Starting bulk user creation for {len(email_list)} email(s)...\n")
    
    # Process emails
    successful = []
    failed = []
    
    for i, email in enumerate(email_list, 1):
        print(f"📧 [{i}/{len(email_list)}] Creating account for: {email}...", end=' ')
        sys.stdout.flush()
        
        result = create_user(email)
        
        if result['success']:
            user_id = result.get('user_id', 'unknown')
            profile_warning = result.get('profile_warning', False)
            status = "✅ Success" if not profile_warning else "⚠️  Partial"
            print(f"{status}: {email} (User ID: {user_id[:8]}...)")
            successful.append({'email': email, 'user_id': user_id, 'warning': profile_warning})
        else:
            error = result.get('error', 'Unknown error')
            print(f"❌ Failed: {email} - {error}")
            failed.append({'email': email, 'error': error})
        
        # Small delay to avoid rate limiting
        time.sleep(0.5)
    
    # Print summary
    print("\n" + "=" * 60)
    print("📊 BULK USER CREATION SUMMARY")
    print("=" * 60)
    print(f"Total emails: {len(email_list)}")
    print(f"✅ Successful: {len(successful)}")
    print(f"❌ Failed: {len(failed)}")
    print()
    
    if successful:
        print("✅ SUCCESSFUL CREATIONS:")
        for item in successful:
            warning_text = " (profile warning)" if item['warning'] else ""
            print(f"   - {item['email']}{warning_text}")
        print()
    
    if failed:
        print("❌ FAILED CREATIONS:")
        for item in failed:
            print(f"   - {item['email']}: {item['error']}")
        print()
    
    print("=" * 60)
    print()
    print(f"📝 Default password for all accounts: \"{DEFAULT_PASSWORD}\"")
    print("📝 Note: Users are created with status \"pending\" and need approval.")
    print()

if __name__ == '__main__':
    main()

